oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/BearerTokenAuthenticationFilter.java

```diff
@@ -0,0 +1,163 @@
+import org.springframework.security.web.AuthenticationEntryPoint;
+import org.springframework.security.web.authentication.WebAuthenticationDetailsSource;
+import org.springframework.util.Assert;
+import org.springframework.web.filter.OncePerRequestFilter;
+
+/**
+ * Authenticates requests that contain an OAuth 2.0
+ * <a href="https://tools.ietf.org/html/rfc6750#section-1.2" target="_blank">Bearer Token</a>.
+ *
+ * This filter should be wired with an {@link AuthenticationManager} that can authenticate a
+ * {@link BearerTokenAuthenticationToken}.
+ *
+ * @author Josh Cummings
+ * @author Vedran Pavic
+ * @author Joe Grandja
+ * @since 5.1
+ * @see <a href="https://tools.ietf.org/html/rfc6750" target="_blank">The OAuth 2.0 Authorization Framework: Bearer Token Usage</a>
+ * @see JwtAuthenticationProvider
+ */
+public final class BearerTokenAuthenticationFilter extends OncePerRequestFilter {
+	private final AuthenticationManager authenticationManager;
+
+	private final AuthenticationDetailsSource<HttpServletRequest, ?> authenticationDetailsSource =
+			new WebAuthenticationDetailsSource();
+
+	private BearerTokenResolver bearerTokenResolver = new DefaultBearerTokenResolver();
+
+	private AuthenticationEntryPoint authenticationEntryPoint = new BearerTokenAuthenticationEntryPoint();
+
+	/**
+	 * Construct a {@code BearerTokenAuthenticationFilter} using the provided parameter(s)
+	 * @param authenticationManager
+	 */
+	public BearerTokenAuthenticationFilter(AuthenticationManager authenticationManager) {
+		Assert.notNull(authenticationManager, "authenticationManager cannot be null");
+		this.authenticationManager = authenticationManager;
+	}
+
+	/**
+	 * Construct a {@code BearerTokenAuthenticationFilter} using the provided parameter(s)
+	 * @param authenticationManager
+	 * @param authenticationEntryPoint
+	 */
+	public BearerTokenAuthenticationFilter(
+			AuthenticationManager authenticationManager,
+			AuthenticationEntryPoint authenticationEntryPoint) {
+
+		this(authenticationManager);
+
+		Assert.notNull(authenticationEntryPoint, "authenticationEntryPoint cannot be null");
+		this.authenticationEntryPoint = authenticationEntryPoint;
+	}
+
+	/**
+	 * Extract any <a href="https://tools.ietf.org/html/rfc6750#section-1.2" target="_blank">Bearer Token</a> from
+	 * the request and attempt an authentication.
+	 *
+	 * @param request
+	 * @param response
+	 * @param filterChain
+	 * @throws ServletException
+	 * @throws IOException
+	 */
+	@Override
+	protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain)
+			throws ServletException, IOException {
+
+		final boolean debug = this.logger.isDebugEnabled();
+
+		String token;
+
+		try {
+			token = this.bearerTokenResolver.resolve(request);
+		} catch ( OAuth2AuthenticationException invalid ) {
+			this.authenticationEntryPoint.commence(request, response, invalid);
+			return;
+		} catch ( Exception unexpected ) {
```

Comment:
Why do we continue processing the filter chain here? I'd expect this fails. If we are expecting this to be a normal flow we should not log it as an error.

Reply:
@rwinch, No, it is not a normal flow.

My reasoning is: If the bearer token resolver fails in an unanticipated way, then we cannot know at that point that it is definitely a bearer token request--it might be another type of authentication. In that circumstance, it seemed like this filter would play nicer with others if it yielded.

The alternative, as I see it, is to fail fast, e.g. not have the extra `Exception` block. The trade-off is that if the bearer token resolver has some kind of bug that throws an unexpected exception, that might limit this filter's ability to coexist with downstream filters (which may have been able to handle the request correctly).

Certainly, bearer token resolution is simple enough that such a circumstance is pretty unlikely, so, I think a less polite posture would also be fine if you see it causing more headaches than it is worth. What is your experience regarding past bugs in other similar filters?

Reply:
Can you clarify what you meant here: 

> Allowing it to pass through makes it confusing as a user would then get some sort of other error that isn't related to the bearer token.

This statement makes sense if our premise is that this request is a bearer token request, and we just messed up in determining that.

My premise, though, is that if an unanticipated exception is thrown in the resolver, then we don't know that it is a bearer token request since the code that determines whether or not this filter wants to participate has a bug itself.

If I don't know that it is a bearer token request, then why should I be worried about errors from downstream filters (indeed those errors may be the correct ones since it may be a request of another type)?

Again, just looking for clarification on that specific point. Fail-fast is a good principle, and it would certainly help us prioritize that (hypothetical) bug instead of it lingering in the codebase.

Reply:
If you continue the FilterChain vs letting the Exception be thrown, it draws attention away from the actual problem. If there is a bug in code, let it have a bug and let the developer know where it is by allowing the Exception to be thrown.

Consider a situation where a client is trying to use a Bearer token, there is a bug, and we continue processing the request after an Exception is thrown. The developer now has much more code to sift through to figure out what is happening since we continue the FilterChain and there is more logging to look through.